util/test/demos/d3d11/d3d11_test.cpp

```diff
@@ -114,65 +114,50 @@ void D3D11GraphicsTest::Prepare(int argc, char **argv)
       ID3D11DevicePtr tempDev;
 
       D3D_FEATURE_LEVEL features[] = {D3D_FEATURE_LEVEL_11_0};
-      hr = dyn_D3D11CreateDevice(adapter, D3D_DRIVER_TYPE_HARDWARE, NULL, 0, features, 1,
-                                 D3D11_SDK_VERSION, &tempDev, NULL, NULL);
+      hr = CreateDevice(adapters, NULL, features, 0);
 
       if(SUCCEEDED(hr))
       {
-        tempDev->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS, &opts, sizeof(opts));
+        dev->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS, &opts, sizeof(opts));
 
         ID3D11Device1Ptr tempDev1;
 
-        tempDev1 = tempDev;
+        tempDev1 = dev;
         memset(&opts1, 0, sizeof(opts1));
         if(tempDev1)
           tempDev1->CheckFeatureSupport(D3D11_FEATURE_D3D11_OPTIONS1, &opts1, sizeof(opts1));
       }
+
+      // This device was only used  to get feature support. Set it back to NULL
+      dev = NULL;
     }
   }
 }
 
-bool D3D11GraphicsTest::Init()
+bool D3D11GraphicsTest::Init(IDXGIAdapter **ppAdapter)
 {
   if(!GraphicsTest::Init())
     return false;
 
   D3D_FEATURE_LEVEL features[] = {D3D_FEATURE_LEVEL_11_0};
-  D3D_DRIVER_TYPE driver = D3D_DRIVER_TYPE_HARDWARE;
 
   if(d3d11_1)
     features[0] = D3D_FEATURE_LEVEL_11_1;
 
-  if(warp)
-    driver = D3D_DRIVER_TYPE_WARP;
-
-  if(adapter)
-    driver = D3D_DRIVER_TYPE_UNKNOWN;
-
   HRESULT hr = S_OK;
 
   UINT flags = createFlags | (debugDevice ? D3D11_CREATE_DEVICE_DEBUG : 0);
 
   if(headless)
   {
-    hr = dyn_D3D11CreateDevice(adapter, driver, NULL, flags, features, 1, D3D11_SDK_VERSION, &dev,
-                               NULL, &ctx);
-
-    // if it failed but on a high feature level, try again on warp
-    if(FAILED(hr) && features[0] != D3D_FEATURE_LEVEL_11_0)
+    if(ppAdapter != NULL)
     {
-      driver = D3D_DRIVER_TYPE_WARP;
-      hr = dyn_D3D11CreateDevice(adapter, driver, NULL, flags, features, 1, D3D11_SDK_VERSION, &dev,
-                                 NULL, &ctx);
-    }
-
-    // if it failed again on a high feature level, try last on ref
-    if(FAILED(hr) && features[0] != D3D_FEATURE_LEVEL_11_0)
-    {
-      driver = D3D_DRIVER_TYPE_REFERENCE;
-      hr = dyn_D3D11CreateDevice(adapter, driver, NULL, flags, features, 1, D3D11_SDK_VERSION, &dev,
-                                 NULL, &ctx);
+      std::vector<IDXGIAdapterPtr> specifiedAdapter;
+      specifiedAdapter.push_back(*ppAdapter);
+      hr = CreateDevice(specifiedAdapter, NULL, features, flags);
```

Comment:
Could be simplified to just `CreateDevice({adapter}, NULL, features, flags);`. Doesn't change the generated code but it's a bit easier to read.